File: web/apps/api/src/scripts/replay-ingest.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { WebSocket } from "ws";

/**
 * End-to-end harness: replay a sim world's frames through the LIVE stack
 * (ingest API -> Redis queue -> Python worker -> Redis pub/sub -> WebSocket) and
 * assert the pushed deltas + final /bays state match the offline result. This is
 * the plan's "ingest -> push loop" verification, run without a real drone.
 *
 *   API_KEY=<key> tsx src/scripts/replay-ingest.ts [world] [apiBase]
 */
const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = join(here, "..", "..", "..", "..", ".."); // scripts->src->api->apps->web->root

const world = process.argv[2] ?? "fmi_block";
const apiBase = process.argv[3] ?? "http://localhost:4000";
const apiKey = process.env.API_KEY;
if (!apiKey) throw new Error("set API_KEY (from register-drone)");

const droneId = process.env.DRONE_ID ?? "drone-1";
const outDir = join(repoRoot, "sim", "output", world);

interface Pose {
  i: number;
  x: number;
  y: number;
  alt: number;
  yaw: number;
  roll: number;
  pitch: number;
  wp: [number, number];
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

async function main(): Promise<void> {
  const poses: Pose[] = JSON.parse(readFileSync(join(outDir, "poses.json"), "utf8"));
  const expected = JSON.parse(
    readFileSync(join(outDir, "occupancy_results.json"), "utf8"),
  ).results as Record<string, { pred: boolean }>;

  // 1) collect pushed deltas over the WebSocket
  const received = new Map<string, boolean>();
  const ws = new WebSocket(`${apiBase.replace("http", "ws")}/ws/occupancy`);
  await new Promise<void>((resolve, reject) => {
    ws.once("open", () => resolve());
    ws.once("error", reject);
  });
  ws.on("message", (buf) => {
    const msg = JSON.parse(buf.toString());
    if (msg.type === "bay_delta") received.set(msg.bay_id, msg.occupied);
  });
  console.log("ws connected");

  // 2) start a mission
  const mission = await postJson(`${apiBase}/api/v1/ingest/mission/start`, {
    world,
    area: "verification",
    frames_expected: poses.length,
  });
  console.log("mission", mission.mission_id);

  // 3) stream every frame through the ingest endpoint
  for (const pose of poses) {
    const png = readFileSync(join(outDir, `frame_${String(pose.i).padStart(3, "0")}.png`));
    const form = new FormData();
    form.append("frame", new Blob([png], { type: "image/png" }), `frame_${pose.i}.png`);
    form.append(
      "meta",
      JSON.stringify({ drone_id: droneId, world, mission_id: mission.mission_id, pose }),
    );
    const res = await fetch(`${apiBase}/api/v1/ingest/frame`, {
      method: "POST",
      headers: { "x-api-key": apiKey! },
      body: form,
    });
    if (res.status !== 202) console.warn(`frame ${pose.i}: HTTP ${res.status}`);
  }
  console.log(`posted ${poses.length} frames`);

  // 4) wait for the worker to drain + deltas to settle. Break only once we've
  //    actually received deltas AND the count has been stable for 2s (avoids
  //    bailing during the initial gap before the first delta arrives).
  let prev = -1;
  let stable = 0;
  for (let t = 0; t < 40; t++) {
    await sleep(1000);
    if (received.size > 0 && received.size === prev) {
      if (++stable >= 2) break;
    } else {
      stable = 0;
    }
    prev = received.size;
  }
  await postJson(`${apiBase}/api/v1/ingest/mission/${mission.mission_id}/end`, {});
  ws.close();

  // 5) assert final /bays state matches the offline predictions
  const fc = (await (await fetch(`${apiBase}/api/v1/bays`)).json()) as {
    features: Array<{ properties: { bay_id: string; occupied: boolean | null } }>;
  };
  const stateById = new Map<string, boolean | null>();
  for (const f of fc.features) {
    stateById.set(String(f.properties.bay_id), f.properties.occupied);
  }
  let match = 0;
  let mismatch = 0;
  const bad: string[] = [];
  for (const [bayId, r] of Object.entries(expected)) {
    const got = stateById.get(bayId);
    if (got === r.pred) match++;
    else {
      mismatch++;
      bad.push(`${bayId}: offline=${r.pred} live=${got}`);
    }
  }

  console.log(`\nWebSocket deltas received: ${received.size}`);
  console.log(`final /bays vs occupancy_results: match=${match} mismatch=${mismatch}`);
  if (bad.length) console.log("  " + bad.slice(0, 20).join("\n  "));
  process.exit(mismatch ? 1 : 0);
}

async function postJson(url: string, body: unknown): Promise<any> {
  const res = await fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json", "x-api-key": apiKey! },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(`POST ${url} -> ${res.status}: ${await res.text()}`);
  return res.json();
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: web/apps/vision-worker/parkdrone_vision/replay_ingest.py
"""End-to-end harness: replay a sim survey area's frames through the LIVE server
(ingest -> in-process queue -> classify threads -> WebSocket push) and assert the
pushed deltas + final /bays state match the offline result. Exercises the same
HTTP/WS contract the drone and the dashboard use.

    API_KEY=<key> python -m parkdrone_vision.replay_ingest [survey_area] [api_base]
"""
import asyncio
import json
import os
import sys
import urllib.error
import urllib.request
import uuid

from .config import SIM_OUTPUT_ROOT
from .vision.scoring import pose_idx


def _post_json(url, obj, api_key=None):
    req = urllib.request.Request(url, data=json.dumps(obj).encode(), method="POST")
    req.add_header("content-type", "application/json")
    if api_key:
        req.add_header("x-api-key", api_key)
    with urllib.request.urlopen(req) as resp:
        return json.loads(resp.read())


def _get_json(url):
    with urllib.request.urlopen(url) as resp:
        return json.loads(resp.read())


def _post_frame(url, api_key, png, meta_str, filename):
    boundary = uuid.uuid4().hex
    body = (
        f'--{boundary}\r\nContent-Disposition: form-data; name="meta"\r\n\r\n'
        f"{meta_str}\r\n"
    ).encode()
    body += (
        f'--{boundary}\r\nContent-Disposition: form-data; name="frame"; '
        f'filename="{filename}"\r\nContent-Type: image/png\r\n\r\n'
    ).encode()
    body += png + b"\r\n" + f"--{boundary}--\r\n".encode()
    req = urllib.request.Request(url, data=body, method="POST")
    req.add_header("Content-Type", f"multipart/form-data; boundary={boundary}")
    req.add_header("x-api-key", api_key)
    try:
        with urllib.request.urlopen(req) as resp:
            return resp.status
    except urllib.error.HTTPError as e:
        return e.code


async def main() -> None:
    survey_area = sys.argv[1] if len(sys.argv) > 1 else "fmi_block"
    api_base = sys.argv[2] if len(sys.argv) > 2 else "http://localhost:4000"
    api_key = os.environ.get("API_KEY")
    if not api_key:
        raise SystemExit("set API_KEY (from register_drone)")
    drone_id = os.environ.get("DRONE_ID", "drone-1")
    out_dir = os.path.join(SIM_OUTPUT_ROOT, survey_area)

    with open(os.path.join(out_dir, "poses.json"), encoding="utf-8") as f:
        poses = json.load(f)
    with open(os.path.join(out_dir, "occupancy_results.json"), encoding="utf-8") as f:
        expected = json.load(f)["results"]

    loop = asyncio.get_running_loop()

    # 1) collect pushed deltas over the WebSocket
    import websockets

    received: dict[str, bool] = {}
    stop = asyncio.Event()

    async def collector():
        async with websockets.connect(api_base.replace("http", "ws") + "/ws/occupancy") as ws:
            while not stop.is_set():
                try:
                    raw = await asyncio.wait_for(ws.recv(), timeout=1.0)
                except asyncio.TimeoutError:
                    continue
                msg = json.loads(raw)
                if msg.get("type") == "bay_delta":
                    received[msg["bay_id"]] = msg["occupied"]

    task = asyncio.create_task(collector())
    await asyncio.sleep(0.5)
    print("ws connected")

    # 2) start a mission
    mission = await loop.run_in_executor(
        None, _post_json, f"{api_base}/api/v1/ingest/mission/start",
        {"survey_area": survey_area, "area": "verification", "frames_expected": len(poses)}, api_key,
    )
    print("mission", mission["mission_id"])

    # 3) stream every frame through the ingest endpoint
    for pose in poses:
        i = pose_idx(pose)
        with open(os.path.join(out_dir, f"frame_{i:03d}.png"), "rb") as f:
            png = f.read()
        meta = json.dumps(
            {"drone_id": drone_id, "survey_area": survey_area, "mission_id": mission["mission_id"], "pose": pose}
        )
        status = await loop.run_in_executor(
            None, _post_frame, f"{api_base}/api/v1/ingest/frame", api_key, png, meta, f"frame_{i}.png"
        )
        if status != 202:
            print(f"frame {i}: HTTP {status}")
    print(f"posted {len(poses)} frames")

    # 4) wait for the queue to drain + deltas to settle (stable for 2s)
    prev, stable = -1, 0
    for _ in range(40):
        await asyncio.sleep(1.0)
        if received and len(received) == prev:
            stable += 1
            if stable >= 2:
                break
        else:
            stable = 0
        prev = len(received)

    await loop.run_in_executor(
        None, _post_json, f"{api_base}/api/v1/ingest/mission/{mission['mission_id']}/end", {}, api_key
    )
    stop.set()
    await task

    # 5) assert final /bays state matches the offline predictions
    fc = await loop.run_in_executor(None, _get_json, f"{api_base}/api/v1/bays")
    state = {str(f["properties"]["bay_id"]): f["properties"]["occupied"] for f in fc["features"]}
    match = mismatch = 0
    bad = []
    for bay_id, r in expected.items():
        if state.get(bay_id) == r["pred"]:
            match += 1
        else:
            mismatch += 1
            bad.append(f"{bay_id}: offline={r['pred']} live={state.get(bay_id)}")

    print(f"\nWebSocket deltas received: {len(received)}")
    print(f"final /bays vs occupancy_results: match={match} mismatch={mismatch}")
    if bad:
        print("  " + "\n  ".join(bad[:20]))
    sys.exit(1 if mismatch else 0)


if __name__ == "__main__":
    asyncio.run(main())
